src/main/java/com/amazonaws/encryptionsdk/keyrings/MultiKeyring.java

```diff
@@ -45,30 +45,34 @@
     }
 
     @Override
-    public void onEncrypt(EncryptionMaterials encryptionMaterials) {
+    public EncryptionMaterials onEncrypt(EncryptionMaterials encryptionMaterials) {
         requireNonNull(encryptionMaterials, "encryptionMaterials are required");
 
+        EncryptionMaterials resultMaterials = encryptionMaterials;
+
         if (generatorKeyring != null) {
-            generatorKeyring.onEncrypt(encryptionMaterials);
+            resultMaterials = generatorKeyring.onEncrypt(encryptionMaterials);
         }
 
-        if (!encryptionMaterials.hasCleartextDataKey()) {
+        if (!resultMaterials.hasCleartextDataKey()) {
             throw new AwsCryptoException("Either a generator keyring must be supplied that produces a cleartext " +
                     "data key or a cleartext data key must already be present in the encryption materials.");
         }
 
         for (Keyring keyring : childrenKeyrings) {
-            keyring.onEncrypt(encryptionMaterials);
+            resultMaterials = keyring.onEncrypt(resultMaterials);
```

Comment:
Not relevant to this particular PR, but are we clear on whether there are any "non-fatal" exceptions Keyrings can throw that would still allow for succeeding on a different child keyring?

Reply:
The spec just says "If the child keyring's OnEncrypt fails, this OnEncrypt MUST also fail." I think a non-fatal exception is currently just handled by keyrings by returning the original encryption materials instead of throwing an actual exception.

Reply:
Ah you're correct, I was thinking of the `onDecrypt` behaviour here.